CRM/Activityreport/Data.php

```diff
@@ -50,12 +161,63 @@ public static function get($offset = 0, $limit = 0, $multiValuesOffset = 0, $sta
       $params['activity_date_time'] = $activityDateFilter;
     }
 
-    $activities = civicrm_api3('Activity', 'get', $params);
+    while ($offset < $total) {
+      if ($offset) {
+        $offset--;
+      }
 
-    $formattedActivities = self::formatResult($activities['values']);
-    $result = self::splitMultiValues($formattedActivities, $offset, $multiValuesOffset);
+      $params['options']['offset'] = $offset;
 
-    return $result;
+      $activities = civicrm_api3('Activity', 'get', $params);
+
+      $formattedActivities = self::formatResult($activities['values']);
+
+      unset($activities);
+
+      while (!empty($formattedActivities)) {
+        $result = self::splitMultiValues($formattedActivities, $offset, $multiValuesOffset);
+
+        if ($result['info']['date'] !== $date) {
+          $page = 0;
+          $date = $result['info']['date'];
+        }
+
+        $cachedCount += self::cachePacket($result['data'], $date, $page++);
+
+        unset($result['data']);
+
+        $formattedActivities = array_slice($formattedActivities, $result['info']['nextOffset'] - $offset, NULL, TRUE);
+
+        $offset = $result['info']['nextOffset'];
+        $multiValuesOffset =  $result['info']['multiValuesOffset'];
+      }
+    }
+
+    return $cachedCount;
```

Comment:
I didn't split the method yet, would like to see your suggestion about it. The main thing is - maybe this method has misleading name (cacheData) - the main thing it does is to walkthrough all Activities so maybe it should be named 'processAllActivities' (or something like that). The caching instruction is at the lowest level of the method so do you think it would be fine to make the method something like:

processAllActivities($startDate, $endDate, $callback)

and then use the $callback at the place of current caching line?

In this case the method would not 'cache' anything, just walk through the data and execute a callback on particular data results. And we could create a separate method calling:

processAllActivities($startDate, $endDate, function() { some caching instruction })

..or something.

Also in meantime maybe I just ask Madura to test current code regarding caching implementation (to check if it meets our expectations in terms of performance) and then we can fit the code into final shape. What do you think?